### ISSUE TITLE
adding Widget support

### ISSUE THREAD

**Issue by pointhi** [searx/searx#60]
Currently I'm work to get and display weather-data like google. I have code it as engine with custom template, but I think, it would an nice idea, to add support for widgets.

As widget, I would define special functions, like displaying weather data, a calculator,... if required. So, It parse the querry, and when required it display special widgets on the result list (on the top).

**Comment by pointhi** [comment]
@asciimoo it would be greate if offline engines support is working soon.

My implementation ideas:
- probably we can create an option `offline=true` and if this option is chosen only one function is calld
- or we define `url=None` and calling dectecting offline-functionality in this way

Furthermore, I think we should think about an master template for widegets. It would't be handable if we create a new .html-file for every wideget which is showing things a little bit different. Nothing which can handle all cases, but most.

**Comment by asciimoo** [comment]
@pointhi, agree, the only thing that holds this issue is an other approach i'm thinking on. Kind of a plugin system, where plugins can attach themselves to specific events like search start/end and manipulate the actual state of the performed search (eg. reorder/add/filter results, rewrite engine/search settings, ..).

**Comment by nogweii** [comment]
It'd be great to be able to integrate DDG's open source instant answers with searx more directly. Unfortunately, it's unlikely to directly embed them in the source code or server process -- the maintenance burden would be too high. I think something that'd be easier to maintain would be a secondary process written in perl with a lightweight API (HTTP on loopback, or something else) exposing the "goodies". (Unfortunately, not every instant answer is available on Github, some of them remain closed source due to licensing agreements with whomever.)
